File: ios/DataSource.swift
import Foundation
import NitroModules

enum DataSource {
  case http(url: URL)
  case file(url: URL)
  case bundle(resource: String, extension: String?)
  case bytes(data: Data)

  static func fromURL(_ url: URL) -> DataSource {
    return url.isFileURL ? .file(url: url) : .http(url: url)
  }

  static func fromURL(string: String) throws -> DataSource {
    guard let url = URL(string: string) else {
      throw DataLoaderError.invalidURL(string)
    }
    return fromURL(url)
  }

  static func bundle(nameWithExtension: String) -> DataSource {
    let name = (nameWithExtension as NSString).deletingPathExtension
    let ext = (nameWithExtension as NSString).pathExtension
    return .bundle(resource: name, extension: ext.isEmpty ? nil : ext)
  }

  static func bytes(from buffer: ArrayBufferHolder) -> DataSource {
    return .bytes(data: buffer.toData(copyIfNeeded: false))
  }

  func createLoader() -> DataLoader {
    switch self {
    case .http:
      return HTTPDataLoader.shared
    case .file:
      return FileDataLoader()
    case .bundle:
      return BundleDataLoader()
    case .bytes:
      return BytesDataLoader()
    }
  }
}

File: android/src/main/java/com/margelo/nitro/rive/DataSource.kt
package com.margelo.nitro.rive

import com.margelo.nitro.core.ArrayBuffer
import java.net.URI

sealed class DataSource {
  data class Http(val url: String) : DataSource()
  data class File(val path: String) : DataSource()
  data class Resource(val name: String) : DataSource()
  class Bytes(val data: ByteArray) : DataSource() {
    companion object {
      fun from(buffer: ArrayBuffer): Bytes {
        val byteBuffer = buffer.getBuffer(false)
        val byteArray = ByteArray(byteBuffer.remaining())
        byteBuffer.get(byteArray)
        return Bytes(byteArray)
      }
    }
  }

  companion object {
    fun fromURL(url: String): DataSource {
      val uri = URI(url)
      return if (uri.scheme == "file") File(uri.path) else Http(url)
    }

    fun resource(nameWithExtension: String): Resource {
      return Resource(nameWithExtension.substringBeforeLast("."))
    }
  }

  fun createLoader(): DataLoader = when (this) {
    is Http -> HTTPDataLoader
    is File -> FileDataLoader
    is Resource -> ResourceDataLoader
    is Bytes -> BytesDataLoader
  }
}
